File: jlox/lox/Scanner.java
package lox;

import static lox.TokenType.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Scanner {

    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start = 0; // points to first char in lexeme being scanned
    private int current = 0; // points at curr char being considered
    private int line = 1; // helps produce tokens that know their location

    // for handling keywords re: identifier's lexeme:
    private static final Map<String, TokenType> keywords;

    static {
        keywords = new HashMap<>();
        keywords.put("and", AND);
        keywords.put("class", CLASS);
        keywords.put("else", ELSE);
        keywords.put("false", FALSE);
        keywords.put("for", FOR);
        keywords.put("fun", FUN);
        keywords.put("if", IF);
        keywords.put("nil", NIL);
        keywords.put("or", OR);
        keywords.put("print", PRINT);
        keywords.put("return", RETURN);
        keywords.put("super", SUPER);
        keywords.put("this", THIS);
        keywords.put("true", TRUE);
        keywords.put("var", VAR);
        keywords.put("while", WHILE);
    }

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (!isAtEnd()) {
            // beginning of next lexeme
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private void scanToken() {
        char c = advance();

        switch (c) {
            case '(':
                addToken(LEFT_PAREN);
                break;
            case ')':
                addToken(RIGHT_PAREN);
                break;
            case '{':
                addToken(LEFT_BRACE);
                break;
            case '}':
                addToken(RIGHT_BRACE);
                break;
            case ',':
                addToken(COMMA);
                break;
            case '.':
                addToken(DOT);
                break;
            case '-':
                addToken(MINUS);
                break;
            case '+':
                addToken(PLUS);
                break;
            case ';':
                addToken(SEMICOLON);
                break;
            case '*':
                addToken(STAR);
                break;
            // THESE cases need to look at the second char (two char operators)
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;
            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;
            case '<':
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;
            case '>':
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;
            // handling of both comments vs. division op
            case '/':
                if (match('/')) {
                    // comment goes until the end of line
                    while (peek() != '\n' && !isAtEnd()) advance();
                    // note: we use the LOOKAHEAD of `peek` to determine end of comment - we do NOT call addToken (the comment is not 'digested' further)
                } else {
                    addToken(SLASH);
                }
                break;
            // 'meaningless chars' (whitespace/newlines) -- reminder without break we can 'fall through' all cases and THEN break
            case ' ':
            case '\r':
            case '\t':
                break;
            // except newlines, we want to increment `line` val
            case '\n':
                line++;
                break;
            case '"':
                string();
                break;
            // for all float point/number instances and unknown chars error handling
            // re: errors-- at lexical level, what if source file has chars Lox doesn't use?
            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    Lox.error(line, "Unexpected char.");
                }
                break;
            // NOTE: erroneous char is still consumed by call in advance
            // - avoids infinite loop ^
            // - we keep scanning, in case other errors we can report all found in 'one go' (no syntax WACK A MOLE)
            // sets HadError -- no code gets EXECUTED
        }
    }

    private void identifier() {
        while (isAlphaNumeric(peek())) advance();

        String text = source.substring(start, current);

        TokenType type = keywords.get(text);
        if (type == null) type = IDENTIFIER;

        addToken(type);
    }

    private void number() {
        while (isDigit(peek())) advance();

        // look for decimal
        if (peek() == '.' && isDigit(peekNext())) {
            // consume '.'
            advance();

            // find all post decimal numeric vals ('fractional' part)
            while (isDigit(peek())) advance();
        }

        addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') line++;
            advance();
        }

        if (isAtEnd()) {
            Lox.error(line, "Unterminated string.");
            return;
        }

        // to closing '"'
        advance();

        // remove quotes, return the token with the additional literal value (needed later steps)
        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    // a 'conditional' version of `advance`
    private boolean match(char expected) {
        if (isAtEnd()) return false;
        if (source.charAt(current) != expected) return false;

        current++;
        return true;
    }

    // for LOOKAHEAD (without consuming char)
    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(current);
    }

    // for looking past decimal point specifically (need '2' places ahead)
    private char peekNext() {
        if (current + 1 >= source.length()) return '\0';
        return source.charAt(current + 1);
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    // checks if all chars consumed
    private boolean isAtEnd() {
        return current >= source.length();
    }

    // consumes next char from source and returns (for input)
    private char advance() {
        return source.charAt(current++);
    }

    // grabs text at curr lexeme and creates new token (for output)
    private void addToken(TokenType type) {
        addToken(type, null);
    }

    // grabs text at curr lexeme and allows for overload to handle tokens with LITERAL vals
    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }
}

File: rlox/my-ast-interpreter/src/scanner.rs
use crate::{Token, TokenLiteral, TokenType};
use std::collections::HashMap;

pub struct Scanner {
    pub source: Vec<char>,
    pub tokens: Vec<Token>,
    pub errors: Vec<String>,
    keywords: HashMap<&'static str, TokenType>,
    start: usize,
    current: usize,
    line: usize,
}

impl Scanner {
    pub fn new(source: String) -> Self {
        Scanner {
            source: source.chars().collect(),
            tokens: vec![],
            errors: vec![],
            start: 0,
            current: 0,
            line: 1,
            keywords: HashMap::from([
                ("and", TokenType::AND),
                ("class", TokenType::CLASS),
                ("else", TokenType::ELSE),
                ("false", TokenType::FALSE),
                ("for", TokenType::FOR),
                ("fun", TokenType::FUN),
                ("if", TokenType::IF),
                ("nil", TokenType::NIL),
                ("or", TokenType::OR),
                ("print", TokenType::PRINT),
                ("return", TokenType::RETURN),
                ("super", TokenType::SUPER),
                ("this", TokenType::THIS),
                ("true", TokenType::TRUE),
                ("var", TokenType::VAR),
                ("while", TokenType::WHILE),
            ]),
        }
    }

    pub fn scan_tokens(mut self) -> (Vec<Token>, Vec<String>) {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token()
        }

        self.tokens
            .push(Token::new(TokenType::EOF, "".to_string(), None, self.line));

        (self.tokens, self.errors)
    }

    pub fn scan_token(&mut self) {
        let c = self.advance();
        match c {
            '(' => self.add_token(TokenType::LEFTPAREN, None),
            ')' => self.add_token(TokenType::RIGHTPAREN, None),
            '{' => self.add_token(TokenType::LEFTBRACE, None),
            '}' => self.add_token(TokenType::RIGHTBRACE, None),
            ',' => self.add_token(TokenType::COMMA, None),
            '.' => self.add_token(TokenType::DOT, None),
            '-' => self.add_token(TokenType::MINUS, None),
            '+' => self.add_token(TokenType::PLUS, None),
            ';' => self.add_token(TokenType::SEMICOLON, None),
            '*' => self.add_token(TokenType::STAR, None),

            '!' => match self.operator_match('=') {
                true => self.add_token(TokenType::NOTEQUAL, None),
                false => self.add_token(TokenType::BANG, None),
            },
            '=' => match self.operator_match('=') {
                true => self.add_token(TokenType::EQUAL, None),
                false => self.add_token(TokenType::ASSIGN, None),
            },
            '<' => match self.operator_match('=') {
                true => self.add_token(TokenType::LESSEQUAL, None),
                false => self.add_token(TokenType::LESS, None),
            },
            '>' => match self.operator_match('=') {
                true => self.add_token(TokenType::GREATEREQUAL, None),
                false => self.add_token(TokenType::GREATER, None),
            },
            '/' => {
                // advance through comment in code
                if self.operator_match('/') {
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else {
                    self.add_token(TokenType::SLASH, None)
                }
            }
            ' ' | '\r' | '\t' => (),
            '\n' => self.line += 1,

            '"' => self.string(),

            c => {
                if c.is_ascii_digit() {
                    self.number();
                } else if c.is_alphabetic() || c == '_' {
                    self.identifier();
                } else {
                    // unknown char
                    self.errors.push(format!(
                        "[line {}] Error: Unexpected character: {}",
                        self.line, c
                    ))
                }
            }
        }
    }

    fn identifier(&mut self) {
        while self.peek().is_alphanumeric() || self.peek() == '_' {
            self.advance();
        }

        let literal = self.source[self.start..self.current]
            .iter()
            .collect::<String>();

        let token_type = self
            .keywords
            .get(literal.as_str())
            .unwrap_or(&TokenType::IDENTIFIER)
            .clone();

        self.add_token(token_type, None);
    }

    fn number(&mut self) {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        if self.peek() == '.' && self.peek_next().is_ascii_digit() {
            self.advance();

            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        let num_str = &self.source[self.start..self.current]
            .iter()
            .collect::<String>();

        if let Ok(num) = num_str.parse::<f64>() {
            self.add_token(TokenType::NUMBER, Some(TokenLiteral::Number(num)));
        } else {
            self.errors.push(format!(
                "[line {}] Error: Invalid number literal: {}",
                self.line, num_str
            ))
        }
    }

    fn string(&mut self) {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            self.errors
                .push(format!("[line {}] Error: Unterminated string.", self.line));
            return;
        }

        self.advance(); // closing '"'

        let literal = self.source[self.start + 1..self.current - 1]
            .iter()
            .collect::<String>();

        self.add_token(TokenType::STRING, Some(TokenLiteral::String(literal)));
    }

    fn operator_match(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.source[self.current] != expected {
            return false;
        }

        self.current += 1;
        true
    }

    fn peek(&self) -> char {
        if let Some(c) = self.source.get(self.current) {
            *c
        } else {
            '\0'
        }
    }

    fn peek_next(&self) -> char {
        if let Some(c) = self.source.get(self.current + 1) {
            *c
        } else {
            '\0'
        }
    }

    fn advance(&mut self) -> char {
        let c = self.source[self.current];
        self.current += 1;
        c
    }

    fn add_token(&mut self, token_type: TokenType, literal: Option<TokenLiteral>) {
        let text = self.source[self.start..self.current]
            .iter()
            .collect::<String>();
        self.tokens
            .push(Token::new(token_type, text, literal, self.line))
    }

    pub fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }
}
